src/core/heavyGraph.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+import {
+  type Edge,
+  type Node,
+  type EdgeAddressT,
+  type NodeAddressT,
+  Graph,
+} from "./graph";
+import {
+  type NodeWeight,
+  type EdgeWeight,
+  type WeightsComparison,
+} from "./weights";
+
+/**
+ * An entry in the array returned by AddressModule.toParts()
+ */
+export type AddressPart = string;
+
+/**
+ * the core structure in the Node weight-resolution tree
+ */
+export type NodeWeightBranch = {
+  +weight?: NodeWeight,
+  +children: Map<AddressPart, NodeWeightBranch>,
+};
+
+/**
+ * the core structure in the Edge weight-resolution tree
+ */
+export type EdgeWeightBranch = {
+  +weight?: EdgeWeight,
+  +children: Map<AddressPart, EdgeWeightBranch>,
+};
+
+/**
+ * The HeavyGraph class is a replacement for the WeightedGraph type. This class
+ * eliminates the need give every node and edge that exists a weight. Rather,
+ * address prefixes can be assigned a weight and when a node or edge weight is
+ * queried, the longest prefix with a weight assigned will have its value
+ * returned.
+ *
+ * Over the longer-term, WeightedGraph should be
+ * refactored out of existence HeavyGraph should completely replace it.
+ */
+export class HeavyGraph {
+  _nodeWeights: NodeWeightBranch;
+  _edgeWeights: EdgeWeightBranch;
+  graph: Graph;
+
+  /**
+   * Set the weight for a given "class" of nodes. Each node that contains the
+   * supplied prefix in its address will potentially return this weight if no
+   * weight is assigned for a larger prefix that the node's address contains.
+   */
+  setNodePrefixWeight(
+    _unused_prefix: NodeAddressT,
+    _unused_w: NodeWeight
+  ): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Set the weights for a given "class" of edges. Each edge that contains the
+   * supplied prefix in its address will potentially return these weights if
+   * none are assigned for a larger prefix that the edge's address contains.
+   */
+  setEdgePrefixWeight(
+    _unused_prefix: EdgeAddressT,
+    _unused_w: NodeWeight
+  ): this {
+    throw new Error("method not implemented");
+  }
+
+  /**
+   * Add a node and optionally assign its weight. If a prefix of the node's
+   * address doesn't have a weight assigned, a weight must be included or an
+   * error will throw.
+   */
```

Comment:
"If a prefix doesn't have a weight assigned, a weight must be included" -- this is a departure from our existing semantics, which treats the `1` as the default node weight and `{1,1}` as the default edge weight. Having it throw an error makes the behavior inconsistent, consider the following two examples:
```js
let wg = new WeightedGraph2();
let node = exampleNode();
// fine
wg.addNode(node);
wg.setNodePrefixWeight(NodeAddress.empty, 2);


wg = new WeightedGraph2();
wg.setNodePrefixWeight(NodeAddress.empty, 2)
// this will unexpectedly throw an error
wg.addNode(node)
```

We switched the order of operations, and the result should be equivalent, but now it errors instead of succeeding.